src/storage/engine/mod.rs

```diff
@@ -33,7 +33,17 @@ pub const TEMP_DIR: &'static str = "";
 const SEEK_BOUND: usize = 30;
 const DEFAULT_TIMEOUT_SECS: u64 = 5;
 
-pub type Callback<T> = Box<FnBox(Result<T>) + Send>;
+pub type Callback<T> = Box<FnBox((CbContext, Result<T>)) + Send>;
+
+pub struct CbContext {
```

Comment:
Why define another `CbContext`? It change the API a lot. We have many context already, add another one just makes things complicated and confusing. I think it's ugly and unnecessary.